crates/oxc_transformer/src/es2015/arrow_functions.rs

```diff
@@ -324,7 +246,76 @@ impl<'a> ArrowFunctions<'a> {
             ));
         }
         let this_var = this_var.as_ref().unwrap();
-        this_var.create_spanned_read_reference(span, ctx)
+        Some(this_var.create_spanned_read_reference(span, ctx))
+    }
+
+    /// Find arrow function we are currently in, if it's between current node, and where `this` is bound.
+    /// Return its `ScopeId`.
+    fn get_arrow_function_scope(ctx: &mut TraverseCtx<'a>) -> Option<ScopeId> {
+        // `this` inside a class resolves to `this` *outside* the class in:
+        // * `extends` clause
+        // * Computed method key
+        // * Computed property key
+        //
+        // ```js
+        // // All these `this` refer to global `this`
+        // class C extends this {
+        //     [this] = 123;
+        //     static [this] = 123;
+        //     [this]() {}
+        //     static [this]() {}
+        // }
+        // ```
+        //
+        // `this` resolves to the class / class instance (i.e. `this` defined *within* the class) in:
+        // * Class method bodies
+        // * Class property bodies
+        // * Class static blocks
+        //
+        // ```js
+        // // All these `this` refer to `this` defined within the class
+        // class C {
+        //     a = this;
+        //     static b = this;
+        //     #c = this;
+        //     d() { this }
+        //     static e() { this }
+        //     #f() { this }
+        //     static { this }
+        // }
+        // ```
+        //
+        // So in this loop, we only exit when we encounter one of the above.
+        for ancestor in ctx.ancestors() {
+            match ancestor {
+                // Top level
+                Ancestor::ProgramBody(_)
+                // Function (includes class method body)
+                | Ancestor::FunctionTypeParameters(_)
+                | Ancestor::FunctionThisParam(_)
+                | Ancestor::FunctionParams(_)
+                | Ancestor::FunctionReturnType(_)
+                | Ancestor::FunctionBody(_)
+                // Class property body 
+                | Ancestor::PropertyDefinitionValue(_)
+                // Class static block
+                | Ancestor::StaticBlockBody(_) => return None,
+                Ancestor::ArrowFunctionExpressionTypeParameters(func) => {
+                    return Some(func.scope_id().get().unwrap())
+                }
+                Ancestor::ArrowFunctionExpressionParams(func) => {
+                    return Some(func.scope_id().get().unwrap())
+                }
+                Ancestor::ArrowFunctionExpressionReturnType(func) => {
+                    return Some(func.scope_id().get().unwrap())
+                }
+                Ancestor::ArrowFunctionExpressionBody(func) => {
+                    return Some(func.scope_id().get().unwrap())
+                }
```

Comment:
@Dunqing Do we need to consider the TS fields here? Is it possible to have a `this` expression in any of these:

* `Function::type_parameters`
* `Function::this_param`
* `Function::return_type`
* `ArrowFunctionExpression::type_parameters`
* `ArrowFunctionExpression::return_type`

?

(again, my poor knowledge of TypeScript is hindering me)

Reply:
No, We don't care about TS fields. In TypeScript, it has a `TSThisType` rather than `ThisExpression`